hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/NewHoodieParquetFileFormat.scala

```diff
@@ -0,0 +1,353 @@
+  private var supportBatchCalled = false
+  private var supportBatchResult = false
+  override def supportBatch(sparkSession: SparkSession, schema: StructType): Boolean = {
+    if (!supportBatchCalled) {
+      supportBatchCalled = true
+      supportBatchResult = !isMOR && super.supportBatch(sparkSession, schema)
+    }
+    supportBatchResult
+  }
+
+  override def buildReaderWithPartitionValues(sparkSession: SparkSession,
+                                              dataSchema: StructType,
+                                              partitionSchema: StructType,
+                                              requiredSchema: StructType,
+                                              filters: Seq[Filter],
+                                              options: Map[String, String],
+                                              hadoopConf: Configuration): PartitionedFile => Iterator[InternalRow] = {
+
+    val outputSchema = StructType(requiredSchema.fields ++ partitionSchema.fields)
+
+    val requiredSchemaWithMandatory = if (!isMOR || MergeOnReadSnapshotRelation.isProjectionCompatible(tableState.value)) {
+      //add mandatory fields to required schema
+      val added: mutable.Buffer[StructField] = mutable.Buffer[StructField]()
+      for (field <- mandatoryFields) {
+        if (requiredSchema.getFieldIndex(field).isEmpty) {
+          val fieldToAdd = dataSchema.fields(dataSchema.getFieldIndex(field).get)
+          added.append(fieldToAdd)
+        }
+      }
+      val addedFields = StructType(added.toArray)
+      StructType(requiredSchema.toArray ++ addedFields.fields)
+    } else {
+      dataSchema
+    }
+
+    val requiredSchemaSplits = requiredSchemaWithMandatory.fields.partition(f => HoodieRecord.HOODIE_META_COLUMNS_WITH_OPERATION.contains(f.name))
+    val requiredMeta = StructType(requiredSchemaSplits._1)
+    val requiredWithoutMeta = StructType(requiredSchemaSplits._2)
+    val needMetaCols = requiredMeta.nonEmpty
+    val needDataCols = requiredWithoutMeta.nonEmpty
+    // note: this is only the output of the bootstrap merge if isMOR. If it is only bootstrap then the
+    // output will just be outputSchema
+    val bootstrapReaderOutput = StructType(requiredMeta.fields ++ requiredWithoutMeta.fields)
+
+    val skeletonReaderAppend = needMetaCols && isBootstrap && !(needDataCols || isMOR) && partitionSchema.nonEmpty
+    val bootstrapBaseAppend = needDataCols && isBootstrap && !isMOR && partitionSchema.nonEmpty
+
+    val (baseFileReader, preMergeBaseFileReader, skeletonReader, bootstrapBaseReader) = buildFileReaders(sparkSession,
+      dataSchema, partitionSchema, requiredSchema, filters, options, hadoopConf, requiredSchemaWithMandatory,
+      requiredWithoutMeta, requiredMeta)
+
+    val broadcastedHadoopConf = sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+    (file: PartitionedFile) => {
+      file.partitionValues match {
+        case broadcast: PartitionFileSliceMapping =>
```

Comment:
So there are 2 things going on here:
the first is that you are seeing 
`isBootstrap && bootstrapFileOpt.isPresent` and `isMOR && logFiles.nonEmpty` the isBootstrap and isMore are not necessary because if the second condition is true then the first has to be true.

The second thing, that I don't think should be changed is 
```
                  (isMOR, logFiles.nonEmpty) match {
                    case (true, true) => buildMergeOnReadIterator(bootstrapIterator, logFiles, filePath.getParent,
                      bootstrapReaderOutput, requiredSchemaWithMandatory, outputSchema, partitionSchema, partitionValues,
                      broadcastedHadoopConf.value.value)
                    case (true, false) => appendPartitionAndProject(bootstrapIterator, bootstrapReaderOutput,
                      partitionSchema, outputSchema, partitionValues)
                    case (false, false) => bootstrapIterator
                    case (false, true) => throw new IllegalStateException("should not be log files if not mor table")
```
The reasoning for this is that 
if it's mor and it has log files then we need to merge the log files, then append the partition path, then project away any mandatory fields for merging (recordkey and precombine) that aren't required

if it's mor and doesn't have log files, then we need to append the partitionpath 

if it's not mor, then the partitionpath has already been appended by the reader itself, so we just return it.

The final edge case to bring up is if  (requiredSchemaWithMandatory.isEmpty) 
That means that it is a df.count() so we just use the base file reader

Reply:
Makes sense.  What I referred to is, we need to revisit all the complexity and see if there're opportunities to unify the differences among COW and MOR (may require changes to iterators, file readers, etc.).

Reply:
COW I think would work right now, because we would never wrap the partition columns with PartitionFileSliceMapping in the FileIndex so it would just hit the iterator to read with the base file reader. I would probably just add an if at the readers are created to just return the base file reader